diff/diff.go

```diff
@@ -0,0 +1,534 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"database/sql"
+	"fmt"
+	"math/rand"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/juju/errors"
+	"github.com/ngaut/log"
+	"github.com/pingcap/tidb-tools/pkg/db"
+	"github.com/pingcap/tidb-tools/diff/util"
+	"github.com/pingcap/tidb/model"
+	"github.com/pingcap/tidb/types"
+)
+
+// Diff contains two sql DB, used for comparing.
+type Diff struct {
+	db1              *sql.DB
+	db2              *sql.DB
+	dbName           string
+	chunkSize        int
+	sample           int
+	checkThreadCount int
+	useRowID         bool
+	tables           []*TableCheckCfg
+	fixSQLFile       *os.File
+	sqlCh            chan string
+	wg               sync.WaitGroup
+}
+
+// NewDiff returns a Diff instance.
+func NewDiff(db1, db2 *sql.DB, cfg *Config) (diff *Diff, err error) {
+	diff = &Diff{
+		db1:              db1,
+		db2:              db2,
+		dbName:           cfg.SourceDBCfg.Name,
+		chunkSize:        cfg.ChunkSize,
+		sample:           cfg.Sample,
+		checkThreadCount: cfg.CheckThreadCount,
+		useRowID:         cfg.UseRowID,
+		tables:           cfg.Tables,
+		sqlCh:            make(chan string),
+	}
+	for _, table := range diff.tables {
+		table.Info, err = pkgdb.GetTableInfoWithRowID(diff.db1, diff.dbName, table.Name, cfg.UseRowID)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		table.Schema = diff.dbName
+	}
+	diff.fixSQLFile, err = os.Create(cfg.FixSQLFile)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return diff, nil
+}
+
+// Equal tests whether two database have same data and schema.
+func (df *Diff) Equal() (equal bool, err error) {
+	defer func() {
+		df.fixSQLFile.Close()
+		df.db1.Close()
+		df.db2.Close()
+	}()
+
+	df.wg.Add(1)
+	go df.WriteSqls()
+
+	equal = true
+	tbls1, err := pkgdb.GetTables(df.db1, df.dbName)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+
+	tbls2, err := pkgdb.GetTables(df.db2, df.dbName)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+
+	eq := equalStrings(tbls1, tbls2)
+	// len(df.tables) == 0 means check all tables
+	if !eq && len(df.tables) == 0 {
+		log.Errorf("show tables get different table. [source db tables] %v [target db tables] %v", tbls1, tbls2)
+		equal = false
```

Comment:
why still continue diff process?

Reply:
return directly

Reply:
I want find the all difference just run once.